File: EmojiPanel/src/main/java/com/binlee/emoji/third/OkHttpEngine.java
package com.binlee.emoji.third;

import androidx.annotation.NonNull;

import com.binlee.emoji.compat.HttpEngine;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import okhttp3.Authenticator;
import okhttp3.Call;
import okhttp3.Headers;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public final class OkHttpEngine extends HttpEngine<Request, Response> {

    private volatile OkHttpClient mClient;
    private Config mConfig;

    private OkHttpClient client() {
        if (mClient == null) {
            synchronized (this) {
                if (mClient == null) {
                    setupClient(null);
                }
            }
        }
        return mClient;
    }

    @Override
    protected void setupClient(Object client) {
        mConfig = createConfig();
        final OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.connectTimeout(mConfig.connectTimeout, TimeUnit.SECONDS);
        builder.writeTimeout(mConfig.writeTimeout, TimeUnit.SECONDS);
        builder.readTimeout(mConfig.readTimeout, TimeUnit.SECONDS);
        builder.proxy(mConfig.proxy);
        final Authenticator authenticator = mConfig.authenticate == null
                ? Authenticator.NONE
                : (route, response) -> adaptRequest(mConfig.authenticate);
        builder.authenticator(authenticator);
        mClient = builder.build();
    }

    @Override
    public BaseResponse request(@NonNull BaseRequest request) throws IOException {
        final okhttp3.Response raw = client().newCall(adaptRequest(request)).execute();
        final BaseResponse response = adaptResponse(raw, request);
        trace("OkHttpEngine", response);
        return response;
    }

    @Override
    public void request(@NonNull BaseRequest request, @NonNull Callback callback) {
        client().newCall(adaptRequest(request)).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                mainHandler().post(() -> callback.onFailure(e));
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull okhttp3.Response raw)
                    throws IOException {
                final BaseResponse response = adaptResponse(raw, request);
                mainHandler().post(() -> callback.onResponse(response));
                trace("OkHttpEngine", response);
            }
        });
    }

    @Override
    protected okhttp3.Request adaptRequest(BaseRequest request) {
        final okhttp3.Request.Builder builder = new okhttp3.Request.Builder();
        builder.url(buildUrl(request));
        builder.headers(toHeaders(request.headers));
        if (GET.equals(request.method)) {
            builder.get();
        } else if (POST.equals(request.method)) {
            builder.post(toRequestBody(request));
        } else if (PUT.equals(request.method)) {
            builder.put(toRequestBody(request));
        } else if (DELETE.equals(request.method)) {
            builder.delete();
        }
        return builder.build();
    }

    private RequestBody toRequestBody(BaseRequest request) {
        final MultipartBody.Builder builder = new MultipartBody.Builder();
        for (String name : request.params.keySet()) {
            final Object value = request.params.get(name);
            if (value instanceof File) {
                // type File
                builder.addPart(MultipartBody.Part.createFormData("files",
                        ((File) value).getName(),
                        RequestBody.create(null, ((File) value)))
                );
            } else {
                // type String
                builder.addPart(MultipartBody.Part.createFormData(name, String.valueOf(value)));
            }
        }
        return builder.build();
    }

    private Headers toHeaders(Map<String, String> headers) {
        if (headers == null) {
            headers = new HashMap<>();
        }
        if (mConfig.userAgent != null) {
            headers.put("User-Agent", mConfig.userAgent);
        }
        return Headers.of(headers);
    }

    @Override
    protected BaseResponse adaptResponse(okhttp3.Response raw, BaseRequest current) throws IOException {
        final BaseResponse response = new BaseResponse();
        response.protocol = raw.protocol().toString();
        response.code = raw.code();
        response.msg = raw.message();
        response.headers = toHeaders(raw.headers());
        if (raw.body() != null) {
            response.data = raw.body().bytes();
        }
        response.current = current;
        return response;
    }

    private Map<String, String> toHeaders(Headers headers) {
        final Map<String, String> ret = new HashMap<>();
        for (int i = 0; i < headers.size(); i++) {
            ret.put(headers.name(i), headers.value(i));
        }
        return ret;
    }
}

File: EmojiPanel/src/main/java/com/binlee/emoji/third/OkHttpEngine.kt
package com.binlee.emoji.third

import com.binlee.emoji.compat.HttpEngine
import okhttp3.*
import okhttp3.Headers.Companion.toHeaders
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit

class OkHttpEngine : HttpEngine<Request, Response>() {

    @Volatile
    private lateinit var mClient: OkHttpClient
    private lateinit var mConfig: Config

    private fun client(): OkHttpClient? {
        if (mClient == null) {
            synchronized(this) {
                if (mClient == null) {
                    setupClient(null)
                }
            }
        }
        return mClient
    }

    override fun setupClient(client: Any?) {
        mConfig = createConfig()
        val builder = OkHttpClient.Builder()
        builder.connectTimeout(mConfig.connectTimeout.toLong(), TimeUnit.SECONDS)
        builder.writeTimeout(mConfig.writeTimeout.toLong(), TimeUnit.SECONDS)
        builder.readTimeout(mConfig.readTimeout.toLong(), TimeUnit.SECONDS)
        builder.proxy(mConfig.proxy)
        val authenticator = if (mConfig.authenticate == null) {
            Authenticator.NONE
        } else {
            object : Authenticator {
                override fun authenticate(route: Route?, response: Response): Request? {
                    return adaptRequest(mConfig.authenticate!!)
                }
            }
        }
        builder.authenticator(authenticator)
        mClient = builder.build()
    }

    @Throws(IOException::class)
    override fun request(request: BaseRequest): BaseResponse {
        val raw = client()!!.newCall(adaptRequest(request)).execute()
        val response = adaptResponse(raw, request)
        trace("OkHttpEngine", response)
        return response
    }

    override fun request(request: BaseRequest, callback: Callback) {
        client()!!.newCall(adaptRequest(request)).enqueue(object : okhttp3.Callback {
            override fun onFailure(call: Call, e: IOException) {
                mainHandler().post { callback.onFailure(e) }
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, raw: Response) {
                val response = adaptResponse(raw, request)
                mainHandler().post { callback.onResponse(response) }
                trace("OkHttpEngine", response)
            }
        })
    }

    override fun adaptRequest(request: BaseRequest): Request {
        val builder = Request.Builder()
        builder.url(buildUrl(request))
        builder.headers(toHeaders(request.headers))
        when (request.method) {
            GET -> {
                builder.get()
            }
            POST -> {
                builder.post(toRequestBody(request))
            }
            PUT -> {
                builder.put(toRequestBody(request))
            }
            DELETE -> {
                builder.delete()
            }
        }
        return builder.build()
    }

    private fun toRequestBody(request: BaseRequest): RequestBody {
        val builder = MultipartBody.Builder()
        for (name in request.params!!.keys) {
            val value = request.params!![name]
            if (value is File) {
                // type File
                builder.addPart(MultipartBody.Part.Companion.createFormData(
                        "files", value.name, value.asRequestBody())
                )
            } else {
                // type String
                builder.addPart(value.toString().toRequestBody())
            }
        }
        return builder.build()
    }

    private fun toHeaders(headers: MutableMap<String, String>): Headers {
        headers["User-Agent"] = mConfig.userAgent
        return headers.toHeaders()
    }

    @Throws(IOException::class)
    override fun adaptResponse(raw: Response, current: BaseRequest): BaseResponse {
        val response = BaseResponse()
        response.protocol = raw.protocol.toString()
        response.code = raw.code
        response.msg = raw.message
        response.headers = toHeaders(raw.headers)
        if (raw.body != null) {
            response.data = raw.body!!.bytes()
        }
        response.current = current
        return response
    }

    private fun toHeaders(headers: Headers): MutableMap<String?, String?> {
        val ret: MutableMap<String?, String?> = HashMap()
        for (i in 0 until headers.size) {
            ret[headers.name(i)] = headers.value(i)
        }
        return ret
    }
}